enable/examples/demo/enable/sierpinski_zoom.py

```diff
@@ -0,0 +1,203 @@
+        """ Recursive method to determine size and location of triangles to
+        add to input path object, based on the current iteration n.
+
+        This method recursively adds to the path object which can be drawn upon
+        completion of the full recursive call.
+        """
+        size = self.base_width/4 * (1/2)**(n - 1)
+        if n == self.iterations:
+            return
+        else:
+            # find top left corners of next 3 inverted triangles relative to
+            # the top left corner of the current one (ie point)
+            #
+            # In the diagram below, X is point, and the ? represent the next
+            # points to make recursive calls at
+            #
+            #       /\
+            #     ?/__\
+            #    X/_\/_\
+            #    /\    /\
+            #  ?/__\ ?/__\
+            #  /_\/_\/_\/_\
+            #
+
+            rel_to_point_locs = (size/2)*np.array([
+                [-1, -SQRT3],
+                [1, SQRT3],
+                [3, -SQRT3]
+            ])
+            # absolute location of those next points
+            abs_points = point + rel_to_point_locs
+
+            for point in abs_points:
+                self.add_triangle_to_path(
+                    path,
+                    point,
+                    size
+                )
+                self.sierpinski(
+                    path, point, n+1
+                )
+
+    def add_triangle_to_path(self, path, point, size):
+        """ Adds an inverted triangle to the input path, of side length size,
+        using the input point as the upper left corner of the triangle.
+        """
+        x, y = point
+        path.move_to(x, y)
+        path.line_to(x + size, y)
+        path.line_to(x + .5 * size, y - (SQRT3 / 2) * size)
+        path.line_to(x, y)
+
+
+class Viewer(HasTraits):
+
+    iterations = Range(0, 'max_iters')
+
+    triangle = Instance(AbstractOverlay)
+
+    comp = Instance(Component)
+
+    base_width = Int(500)
+
+    max_iters = Property(observe="base_width, viewport.zoom")
+
+    viewport = Instance(Viewport)
+
+    def _get_bounds(self):
+        return [self.base_width, self.base_width*(SQRT3/2)]
+
+    def _get_max_iters(self):
+        factor = 1
+        if self.viewport:
+            factor = self.viewport.zoom
+        return int(np.log(4/(factor*self.base_width))/np.log(.5) + 1)
+
+    def _comp_default(self):
+        canvas = Canvas(draw_axes=True, bgcolor="gray")
+        self.triangle = SierpinskiTriangle(
+            canvas,
+            position=[0.0, 0.0],
+            bounds=[self.base_width, self.base_width*(SQRT3/2)],
+            iterations=self.iterations,
+            max_iters=self.max_iters,
+            base_width=self.base_width,
+        )
+        canvas.overlays.append(self.triangle)
+
+        self.viewport = Viewport(
+            component=canvas, enable_zoom=True, stay_inside=True
+        )
+        self.viewport.zoom_tool.min_zoom = 1.0
```

Comment:
this works as I want it to initially (eg when base_width is 500), in that you can not zoom out passed when you see the full triangle.  However, if base_width is changed to 1000 for example now you can not zoom out enough.  

I want to only be able to zoom out as far as bounds.  I have had trouble understanding / getting things to work with the `stay_inside` trait, `view_bounds`, etc.

Reply:
Not sure how much I can help. I have basically no experience with `Viewport`.

Reply:
No worries.
I am curious if this (viewport, canvas, etc.) gets much (_any_) use out the wild